### PR TITLE
⬆️ zb: Swtich to new smol-rs crates

### LINKED ISSUE #479: Bump to smol-rs subcrates' 3.0
Hopefully it's not a lot of work.

### PR DESCRIPTION
**Pull request by MaxVerevkin**
closes #479.

### CODE REVIEW

File: zbus/src/connection/mod.rs [hunk 1]
```diff
@@ -1205,7 +1205,7 @@ impl Connection {
     /// Returns a listener, notified on various connection activity.
     ///
     /// This function is meant for the caller to implement idle or timeout on inactivity.
-    pub fn monitor_activity(&self) -> EventListener {
+    pub fn monitor_activity(&self) -> Pin<Box<EventListener>> {
```
[the remark this answers is not in the document]
Comment (MaxVerevkin) [reply zbus/src/connection/mod.rs:1208]: Yeah, this is public-only API. I tried this but now a test timeouts

```
---- blocking::connection::tests::unix_p2p stdout ----
thread 'blocking::connection::tests::unix_p2p' panicked at zbus/src/blocking/connection/mod.rs:292:5:
timeout: the function call took 15000 ms. Max time 15000 ms
```

```diff
diff --git a/zbus/src/blocking/connection/mod.rs b/zbus/src/blocking/connection/mod.rs
index ff4209d5..859d45f3 100644
--- a/zbus/src/blocking/connection/mod.rs
+++ b/zbus/src/blocking/connection/mod.rs
@@ -3,7 +3,7 @@
 use enumflags2::BitFlags;
 use event_listener::EventListener;
 use static_assertions::assert_impl_all;
-use std::{io, ops::Deref, pin::Pin};
+use std::{io, ops::Deref};
 use zbus_names::{BusName, ErrorName, InterfaceName, MemberName, OwnedUniqueName, WellKnownName};
 use zvariant::ObjectPath;
 
@@ -241,7 +241,7 @@ impl Connection {
     /// Returns a listener, notified on various connection activity.
     ///
     /// This function is meant for the caller to implement idle or timeout on inactivity.
-    pub fn monitor_activity(&self) -> Pin<Box<EventListener>> {
+    pub fn monitor_activity(&self) -> EventListener {
         self.inner.monitor_activity()
     }
 
@@ -313,7 +313,7 @@ mod tests {
         });
 
         let c = Builder::unix_stream(p1).p2p().build().unwrap();
-        let mut listener = c.monitor_activity();
+        let mut listener = Box::pin(c.monitor_activity());
         let mut s = MessageIterator::from(&c);
         tx.send(()).unwrap();
         let m = s.next().unwrap().unwrap();
@@ -326,10 +326,12 @@ mod tests {
         assert_eq!(val, "yay");
 
         // there was some activity
+        listener.as_mut().listen();
         listener.as_mut().wait();
         // eventually, nothing happens and it will timeout
         loop {
-            let mut listener = c.monitor_activity();
+            let mut listener = Box::pin(c.monitor_activity());
+            listener.as_mut().listen();
             if listener
                 .as_mut()
                 .wait_timeout(std::time::Duration::from_millis(10))
diff --git a/zbus/src/connection/mod.rs b/zbus/src/connection/mod.rs
index ac5a6352..55da7181 100644
--- a/zbus/src/connection/mod.rs
+++ b/zbus/src/connection/mod.rs
@@ -1205,8 +1205,8 @@ impl Connection {
     /// Returns a listener, notified on various connection activity.
     ///
     /// This function is meant for the caller to implement idle or timeout on inactivity.
-    pub fn monitor_activity(&self) -> Pin<Box<EventListener>> {
-        self.inner.activity_event.listen()
+    pub fn monitor_activity(&self) -> EventListener {
+        EventListener::new(&self.inner.activity_event)
     }
 
     /// Returns the peer credentials.
```

**Comment by MaxVerevkin** [comment]
> You may want to refer to #479 in the description btw.

There aren't any other outdated smol crates, right?

File: zbus/src/connection/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (zeenix) [reply zbus/src/connection/mod.rs:1208]: > Yeah, this is public-only API. I tried this but now a test timeouts

That's very strange. I don't see how they can be connected? Perhaps it's just some timing issue?

**Comment by zeenix** [comment]
> Maybe not at the moment but all crates are being updated.

E.g https://github.com/smol-rs/async-io/pull/155

**Comment by MaxVerevkin** [comment]
> I should roll out a release soon..

Didn't know you were involved in smol-rs :)

File: zbus/src/connection/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (MaxVerevkin) [reply zbus/src/connection/mod.rs:1208]: > Perhaps it's just some timing issue?

Thanks for the idea. Yes! `listener.as_mut().listen();` should be called earlier.

[the remark this answers is not in the document]
Comment (MaxVerevkin) [reply zbus/src/connection/mod.rs:1208]: Now different tests timeout...

```
---- iface_and_proxy_unix_p2p stdout ----
thread 'zbus::Connection executor' panicked at /home/maxv/.cargo/registry/src/index.crates.io-6f17d22bba15001f/event-listener-3.0.0/src/lib.rs:1069:36:
listener was never inserted into the list
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'iface_and_proxy_unix_p2p' panicked at zbus/tests/e2e.rs:688:1:
timeout: the function call took 15000 ms. Max time 15000 ms

---- iface_and_proxy stdout ----
thread 'zbus::Connection executor' panicked at /home/maxv/.cargo/registry/src/index.crates.io-6f17d22bba15001f/event-listener-3.0.0/src/lib.rs:1069:36:
listener was never inserted into the list
thread 'iface_and_proxy' panicked at zbus/tests/e2e.rs:681:1:
timeout: the function call took 15000 ms. Max time 15000 ms
```

**Comment by zeenix** [comment]
> > I should roll out a release soon..
> 
> Didn't know you were involved in smol-rs :)

Not as much as I should be or want to be. :) async-broadcast 0.6.0 now out btw.

File: zbus/src/connection/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (zeenix) [reply zbus/src/connection/mod.rs:1208]: > Now different tests timeout...

:(

I'll look later this week if you don't manage to figure it out by then. BTW, I'd suggest first running tests locally to get a quicker feedback.

[the remark this answers is not in the document]
Comment (MaxVerevkin) [reply zbus/src/connection/mod.rs:1208]: I do run them locally. I will push current (broken) code though so you can reproduce the timeout.

**Comment by MaxVerevkin** [comment]
I don't understand how CI successfully completed on everything except `windows_test`. It certainly doesn't work on my linux.

**Comment by zeenix** [comment]
> Closes #479

As is, it does not closes #479 fully. The issue covers all smol-rs bumps that will be needed.

**Comment by zeenix** [comment]
> I don't understand how CI successfully completed on everything except `windows_test`. It certainly doesn't work on my linux.

Gives you a clue at least that it's some timing issue? Also, better this than the other way around. Locally reproducible issues are **much** easier to debug/solve.

**Comment by MaxVerevkin** [comment]
Something's wrong with the tests, even on master... I remember they used to pass, but now `fdpass_systemd` consistently fails and `unix_p2p` _inconsistently_ timeouts :(

```
---- tests::fdpass_systemd stdout ----
thread '<unnamed>' panicked at zbus/src/lib.rs:323:14:
called `Result::unwrap()` on an `Err` value: MethodError(OwnedErrorName("org.freedesktop.DBus.Error.InteractiveAuthorizationRequired"), Some("Interactive authentication required."), Msg { type: Error, sender: UniqueName(":1.1"), reply-serial: 38, body: Signature("s") })
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'tests::fdpass_systemd' panicked at zbus/src/lib.rs:311:5:
explicit panic

---- connection::tests::unix_p2p stdout ----
thread 'connection::tests::unix_p2p' panicked at zbus/src/connection/mod.rs:1437:5:
timeout: the function call took 15000 ms. Max time 15000 ms
```

**Comment by zeenix** [comment]
> Something's wrong with the tests, even on master... I remember they used to pass, but now `fdpass_systemd` consistently fails and `unix_p2p` _inconsistently_ timeouts :(

Ah yes, this one I've seen happening on my machine too. It disappears after a while though. I'd suggest skipping it as it shouldn't be affected by this PR.

**Comment by zeenix** [comment]
@MaxVerevkin seems like [an event-listener regression](thread 'zbus::Connection executor' panicked at C:\Users\runneradmin\.cargo\registry\src\index.crates.io-6f17d22bba15001f\event-listener-3.0.0\src\lib.rs:1069:36:
listener was never inserted into the list) on Windows:

> thread 'zbus::Connection executor' panicked at C:\Users\runneradmin\.cargo\registry\src\index.crates.io-6f17d22bba15001f\event-listener-3.0.0\src\lib.rs:1069:36:
listener was never inserted into the list

**Comment by zeenix** [comment]
@MaxVerevkin @notgull pointed out that the `EventListener::new` doesn't setup the listener to listen to the given event but rather must be told to do so using `listen` method. I think the new API is very unintuitive and [hopefully we'll have some sort of resolution in event-listener](https://github.com/smol-rs/event-listener/issues/89) for it but it gives you a clue of what must be wrong here.

**Comment by MaxVerevkin** [comment]
> the `EventListener::new` doesn't setup the listener to listen to the given event but rather must be told to do so using `listen` method. I think the new API is very unintuitive and [hopefully we'll have some sort of resolution in event-listener](https://github.com/smol-rs/event-listener/issues/89) for it but it gives you a clue of what must be wrong here.

Even if `.await` is used? Apparently yes, that's unfortunate. I've finally found the problem, just had to run `cargo test -- --nocapture` to see the manic message from a different thread.

**Comment by zeenix** [comment]
> Even if `.await` is used?

Yes because even though `EventListener::new` takes a reference to an `Event`, it can't already set itself to listen to it because that requires a pinned listener. It's bad, I know! BTW do join our matrix channels for discussions (zbus is currently at 99 members so you can set a record :)).

https://matrix.to/#/#zbus:matrix.org
https://matrix.to/#/#smol-rs:matrix.org

**Comment by zeenix** [comment]
Thanks for fixing! One last reservation though: If we only bump our direct event-listener dep, we'll now end up with 2 event-listener versions in our deps, the other through other smol-rs deps. Perhaps it'd make sense to resolve #479 completely in one PR?

**Comment by MaxVerevkin** [comment]
> it can't already set itself to listen to it because that requires a pinned listener.

Can't it do so the first time it is polled?

**Comment by zeenix** [comment]
> > it can't already set itself to listen to it because that requires a pinned listener.
> 
> Can't it do so the first time it is polled?

Perhaps, please comment on https://github.com/smol-rs/event-listener/issues/89

File: zbus/tests/e2e.rs [hunk 2]
```diff
@@ -773,6 +772,7 @@ async fn iface_and_proxy_(p2p: bool) {
     debug!("Client connection created: {:?}", client_conn);
     debug!("Service connection created: {:?}", service_conn);
 
+    let listen = event.listen();
```
Comment (zeenix) [inline zbus/tests/e2e.rs:775]: Why this change? Is it relevant to the update? If not, please split into separate commit with its own log/justification.

**Review by zeenix — CHANGES_REQUESTED** [review]
Nice! Great to see you going the extra mile. :+1: Apart from the nit:

* Just like in case of event-hander, async-broadcast now provides new API to avoid allocation when sending. We should either try to switch to that already here or leave a `FIXME` comment in the code and commit message about that.
* We still have a bunch more crates I think: async-io (it was just released I believe) and async-process I think.

File: zbus/tests/e2e.rs (hunk 2 above)
Comment (MaxVerevkin) [reply zbus/tests/e2e.rs:775]: I was reverting a part of the first commit, but it ended up in the wrong commit

**Comment by MaxVerevkin** [comment]
> Just like in case of event-hander, async-broadcast now provides new API to avoid allocation when sending

I found only one place where this can be done. Done. By the way, the changelog didn't mention this at all, so...

**Comment by MaxVerevkin** [comment]
I did not update async-io because it involves unsafe code, which I would like to avoid in this PR.

**Comment by zeenix** [comment]
> I did not update async-io because it involves unsafe code, which I would like to avoid in this PR.

I don't think that's a good reason to split the whole PR (it's the exactly same topic/general change) but if you prefer it that way, sure. In that case, please do edit the description again to not get #479 closed with this merged.

**Comment by zeenix** [comment]
> I did not update async-io because it involves unsafe code, which I would like to avoid in this PR.

Hm.. would this mean we'll be depending on multiple versions of event-listener? If so, I'd have to go back on my word and ask you to already port to async-io in this PR. Otherwise, we can merge already.

File: zbus/Cargo.toml [hunk 3]
```diff
@@ -108,7 +108,7 @@ nix = { version = "0.27", default-features = false, features = [
 [target.'cfg(target_os = "macos")'.dependencies]
 # FIXME: This should only be enabled if async-io feature is enabled but currently
 # Cargo doesn't provide a way to do that for only specific target OS: https://github.com/rust-lang/cargo/issues/1197.
-async-process = "1.7.0"
+async-process = "2.0.0"
```
Comment (MaxVerevkin) [inline zbus/Cargo.toml:111]: Did not and can not test this change.

Comment (zeenix) [reply zbus/Cargo.toml:111]: Me neither. :) I think some tests cover this so CI will tell us.

**Comment by zeenix** [comment]
@MaxVerevkin Hey, event-listener 4 was released that removes the footgun you encountered and its dependent smol-rs crates have been updated too. Do you think you'll be able to finish this PR soon with those updates?

**Comment by MaxVerevkin** [comment]
> Hey, event-listener 4 was released that removes the footgun you encountered

Thanks for notifying, will look into it!

> Do you think you'll be able to finish this PR soon with those updates?

I hope :) I'll have some time later today.

**Comment by MaxVerevkin** [comment]
@zeenix what do you think `Connection::monitor_activity` should do? Return `Pin<Box<EventListener>>`, `Arc<Event>` or `&Event`?

**Comment by zeenix** [comment]
> @zeenix what do you think `Connection::monitor_activity` should do? Return `Pin<Box<EventListener>>`, `Arc<Event>` or `&Event`?

If our examples work with `EventListener` only then we return that and let the user pin it if they need to. Otherwise, we should return `Pin<Box<EventListener>>`.

**Comment by notgull** [comment]
Just a heads up, `async-lock` and `async-channel` don't use `event-listener` v4 yet. They still need a release. Usually I do a release run on the weekend.

**Comment by MaxVerevkin** [comment]
> If our examples work with EventListener only then we return that

Return what exactly? Returning unpinned EventListener is pointless but returning pinned listener defeats the improvements of `event-listener`.

**Comment by zeenix** [comment]
> > If our examples work with EventListener only then we return that
> 
> Return what exactly? Returning unpinned EventListener is pointless

How so? You can .await it, can't you? If it was a useless type on its own, the new API of `event-listener` wouldn't make any sense.

> but returning pinned listener defeats the improvements of `event-listener`.

Yeah, hence the preference to return `EventListener`.

**Comment by zeenix** [comment]
> Just a heads up, `async-lock` and `async-channel` don't use `event-listener` v4 yet. They still need a release. Usually I do a release run on the weekend.

Thanks, I got ahead of myself. :)

**Comment by MaxVerevkin** [comment]
> > Return what exactly? Returning unpinned EventListener is pointless
> 
> How so?

```rust
pub fn monitor_activity(&self) -> EventListener {
    EventListener::new()
}
```

**Comment by MaxVerevkin** [comment]
Do you want me to update event-listener in a new commit or override the previous one?

**Comment by zeenix** [comment]
> ```rust
> pub fn monitor_activity(&self) -> EventListener {
>     EventListener::new()
> }
> ```

Ah right, I forgot the exact semantics of the weird new API. :( Yes, I think we have to go with `Pin<Box<EventListener>>` then. I don't imagine many users to use more than one listener in their whole application so the allocation is minimal and not something to care about IMO.

> Do you want me to update event-listener in a new commit or override the previous one?

No point in keeping the current one if we're going to change it in the same PR. :) So the latter.

**Comment by MaxVerevkin** [comment]
> No point in keeping the current one if we're going to change it in the same PR. :) So the latter.

Is it ok to leave both? I don't want to mess up the entire git history :laughing:

**Comment by MaxVerevkin** [comment]
> ```
>  thread 'tests::basic_connection_async' panicked at zbus/src/lib.rs:271:5:
> explicit panic
> ```

Ntot sure whats going on in the CI...

**Comment by zeenix** [comment]
> Is it ok to leave both? I don't want to mess up the entire git history 😆

You won't be messing up anything, only minimizing commits but sure, no biggie if you don't want to do it.

> Ntot sure whats going on in the CI...

Well you have to debug obviously. There is some error from the low-level with async-io: https://github.com/dbus2/zbus/actions/runs/6988036119/job/19015207530?pr=494#step:6:705

**Comment by MaxVerevkin** [comment]
I don't know anything about low-level workings of zbus+async-io and honestly don't really want to dig in that direction. I think I'm not the right person to fix this :(

**Comment by zeenix** [comment]
> I don't know anything about low-level workings of zbus+async-io

AFAICT the main errors were from the compiler so I'm sure there were much lowlevel stuff involved.

> I think I'm not the right person to fix this :(

You came so close so it's sad to see you give up on what seems to be last hurdle w/o trying. Of course, it's your decision in the end. Just that I spent time reviewing too. Hopefully that helps me complete the work though.
